rpe/resources/gcp.py

```diff
@@ -486,6 +486,9 @@ class GcpCloudFunction(GoogleAPIResource):
     resource_components = {
         'iam': 'getIamPolicy',
     }
+    readiness_key = 'status'
+    readiness_value = 'ACTIVE'
+    readiness_terminal_values = ['CLOUD_FUNCTION_STATUS_UNSPECIFIED','OFFLINE','UNKNOWN']
```

Comment:
Do the other states prevent us from making changes to the resource?

Reply:
I  just tested a function in the state `OFFLINE` and I was able to patch the function. This is the state of deploying a configured resource. The `UNKNOWN` state say:

> Function deployment failed and the function serving state is undefined. The function should be updated or deleted to move it out of this state.

So it sounds like that supports changes as well.

Reply:
I did some more testing, i got it into the `OFFLINE` state by accident... the cloudbuild api was disabled in my project when I tried to deploy the function, so it failed and went to that state. It allowed me to change the `httpsTrigger` to `secure_always` in that state.

Afterwards, I enabled the cloudbuild API, and redeployed it. While it was deploying the cloud console showed the "deploying" indicator, but the API still returned `OFFLINE` as the status (It never entered the state `DEPLOY_IN_PROGRESS` as i repeated queried the API during deployment). While it was attempting to deploy I got the following error while trying to make the same change as before (I had disabled the above setting between tests):

```
googleapiclient.errors.HttpError: <HttpError 400 when requesting https://cloudfunctions.googleapis.com/v1/projects/xxx/locations/us-central1/functions/function-1?updateMask=httpsTrigger&alt=json returned "An operation on function function-1 in region us-central1 in project xxx is already in progress. Please try again later.". Details: "An operation on function function-1 in region us-central1 in project xxx is already in progress. Please try again later.">
```

Lets discuss this tomorrow and decide how to proceed.

Reply:
I spoke to @BrettPowell , on the creation of a cloudfunction, we're unable to remediate anything when its deploying. So a waiter is exactly the way to go here.